### DISCUSSION TITLE
What is the long-term vision for C# in Godot?

### DISCUSSION THREAD

**Discussion by konstantinkopka** [godotengine/godot-proposals#4604] [Scripting]
Currently we have GDScript as the official, integrated language for scripting in Godot - It's quite intuitive and seems to work well for many users.

However, there are lot of things that people desire that GDScript does not and might never offer: Strong, static typing, external tooling/proper IDEs, real refactoring possibilities, and a mature library ecosystem used outside of Godot (especially important for non-gamedev projects).

C++ modules are an alternative to GDScript, but the workflow is completely different and basically completely bypasses the editor. It's good for very specific, usually performance-oriented problems but not for writing general gameplay code.

For developers wanting a good middle-ground, I thought, C# support was added a few years ago.
I am not overly zealous about C# per se, but I think the language works really well for game development and scripting - I think it's a great fit for Godot and from the last community polling, I think about 20% of developers seem to already use mainly C# with Godot, which I find a lot.

C# is currently the language that's closest to "first-party-supported" as it gets: It is listed alongside GDScript in the docs, there are quite in-depth sections about using it with Godot and it's the only language that gets out of the box support by virtue of the Mono Editor build.

I am aware that currently @neikeq (and @vnen) are working on transitioning over to .NET 6 from the Mono runtime, which I love.

However, and that's the point of this thread: **I feel very unclear about where the majority of the Godot Core Team stands on what role C# should play in Godot's future.**

@YuriSizov wrote in #2430 that
> C# support, while beneficial to some users, is not something that would put the next version of Godot on hold. If neikeq doesn't have time or desire to work on it for 4.0, we'd still release 4.0, without Mono/.NET.

I don't know if this is the majority opinion here but it makes me really worried about building my own, potentially many-year future projects with Godot. It makes a *huge* difference to me if C# is "officially supported" in Godot or has the same status as any other language such as Rust or Nim, that can be used with native bindings.

If 4.0 would really ship without C#, as suggested as a possibility here, this would mean that 20% of current devs could then not use the latest Godot with their chosen language, at all, for possibly quite some time - it would be a disaster.

I do not want to depend on third-party bindings, I need stable and reliable integrations that allow me to focus on my game, and I'm worried that C# might even be "dropped", at least officially, by Godot at some point and relegated to the same position as basically "any other language".

My opinion is that it would be great if Godot actually doubled down on supporting C# from 4.0 onwards - it's a great opportunity for engine growth as well. There is a quickly growing migration from disappointed Unity developers (myself included) that have worked with C# for many years. Attracting those would be amazing for Godot imho, but it can only be done if the core team sees this as a focus area.

As it stands now, I kind of get the feeling that (at least some of) the Godot Core devs view implementing C# as some sort of unfortunate accident, better to be forgotten. I truly hope this is not at all the case, though.

Maybe @reduz can shed some light on what the long-term vision here is?

**Comment by konstantinkopka** [comment]
@aaronfranke has alerted me to the fact that a huge issue is the fact that currently, the effort to port Godot to .NET 6 has no funding.
I have tried reaching out to Scott Hanselman (@shanselman) (Very enthusiastic about improving .NET and works for Microsoft), maybe someone on the core team also has an idea on who to contact at Microsoft? I am sure they could spare 20-40k USD or so to fund this, they're probably just not aware that this is important for the future of C# in Godot.

They are even proudly advertising Godot on their main game engine portal site: https://dotnet.microsoft.com/en-us/apps/games/engines

Maybe we can make something happen?

**Comment by GeorgeS2019** [comment]
@Kopkako neikeq did two commits last 2 days in the dotnet6 branch!

**Comment by reduz** [comment]
Godot as an open source project in nature has a simple logic for these things: For something to happen, somebody has to want to work on it and others must agree that is wanted. 

C# support is clearly wanted and there are contributors working on it. It definitely could be better funded but we failed to get better funding from Microsoft the past years. We will try again this one later in the year.

**Reply by konstantinkopka** [reply]
Thanks, Juan for taking the time to reply!
I am glad you agree that C# support is clearly wanted.

Unity has shifted priorities so many times in the past few years that I feel completely burned out on the engine and do not want to repeat that mistake of committing to something that does not fit.

I aim to be mostly aligned with the general direction Godot is moving towards – to use it more reliably as a platform for building long-term projects myself.

I also really hope Microsoft recognizes the win-win situation for funding Godot - it helps Godot, .NET/C# and makes them look great, for very little cost.

**Comment by konstantinkopka** [comment]
I actually managed to start a bit of a thread on Twitter regarding MS funding possibilities: https://twitter.com/ghostbutter/status/1531096815797383169?s=20&t=7gzhovaGBXQ0Hz1kK3O32w

@reduz Hope that was okay/not overstepping any boundaries - I felt like contacting Scott directly would be a good idea since he is very active in advocating for .NET and I had recently watched a bunch of his videos. Maybe something good can come from this?

**Reply by lentsius-bark** [reply]
Doing god's work!

**Comment by neikeq** [comment]
I can't answer the main questions as they're not directed at me, but I'm going to clarify two things:

- C# with .NET 6 is well on its way to be included in 4.0. The only part that will be  missing is mobile support, as I want to do things differently to 3.x, which will take more time.
- While funding from Microsoft ended some time ago, I'm still being paid for my current work. For this to continue we will likely need to get some funding again, though. In any case, many changes in 4.0 are directed towards making C# easier to maintain and to reduce our workload. So even if at some point no one is working on it full-time, it likely would't fall behind.

**Reply by paulloz** [reply]
> many changes in 4.0 are directed towards making C# easier to maintain and to reduce our workload

This is really cool to hear!  

I'd really be interested in knowing where time from other contributors would be well spent. Most of my contributions to C# in the engine (and, to be fair, to the engine in general) have been small fixes of bugs I either experienced myself or heard of by fellow game devs. And that's mostly due to the fact, coming from the outside, it's really hard to know what would be constructive short/mid/long term. If you have a couple of minutes to talk about that at some point, feel free to @ me on Rocket ❤️

**Reply by konstantinkopka** [reply]
Thanks, @neikeq! I always really enjoy your write-ups on C# development, be it in blog form or in the #2333 issue.
It's great to hear that the future of Godot and using C# is looking bright, really looking forward!

I agree with @paulloz , if there is anything other contributors can do to help, it would be great if there was a place where that's best discussed. Ist Rocket.chat working best for you?

A few interested developers (including Paul and myself) have recently started a (still really small) Discord community around discussing Godot and C# development in particular.

Please (everyone) feel free to join us: https://discord.gg/Qme8nk6aHT
It would be a pleasure to chat with you directly & more casually about C# and Godot, if you're interested.

**Reply by YuriSizov** [reply]
Rocket.Chat is where all engine development discussions happen, yes. We prefer to keep one platform for contributor coordination (plus GitHub, obviously).
